apiserver/cloud/instance_information.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cloud
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/state/stateenvirons"
+)
+
+// EnvironConfigGetter implements environs.EnvironConfigGetter
+// in terms of a *state.State.
+type cloudEnvironConfigGetter struct {
+	Backend
+	region string
+}
+
+// CloudSpec implements environs.EnvironConfigGetter.
+func (g cloudEnvironConfigGetter) CloudSpec(tag names.ModelTag) (environs.CloudSpec, error) {
+	model, err := g.GetModel(tag)
+	if err != nil {
+		return environs.CloudSpec{}, errors.Trace(err)
+	}
+	cloudName := model.Cloud()
+	regionName := g.region
+	credentialTag, _ := model.CloudCredential()
+	return stateenvirons.CloudSpec(g.Backend, cloudName, regionName, credentialTag)
+}
+
+// InstanceTypes returns instance type information for the cloud and region
+// in which the current model is deployed.
+func (api *CloudAPI) InstanceTypes(cons params.CloudInstanceTypesConstraints) (params.InstanceTypesResults, error) {
+	return instanceTypes(api, environs.GetEnviron, cons)
+}
+
+type environGetFunc func(st environs.EnvironConfigGetter, newEnviron environs.NewEnvironFunc) (environs.Environ, error)
+
+func instanceTypes(api *CloudAPI,
+	environGet environGetFunc,
+	cons params.CloudInstanceTypesConstraints,
+) (params.InstanceTypesResults, error) {
+	c, err := api.backend.ControllerModel()
+	if err != nil {
+		return params.InstanceTypesResults{}, errors.Trace(err)
+	}
+	modelTag := c.ModelTag()
+	m, err := api.backend.GetModel(modelTag)
+	if err != nil {
+		return params.InstanceTypesResults{}, errors.Trace(err)
+	}
+
+	result := make([]params.InstanceTypesResult, len(cons.Constraints))
+	// TODO(perrito666) Cache the results to avoid excessive querying of the cloud.
+	for i, cons := range cons.Constraints {
+		value := constraints.Value{}
+		if cons.Constraints != nil {
+			value = *cons.Constraints
+		}
+		backend := cloudEnvironConfigGetter{
+			Backend: api.backend,
+			region:  cons.CloudRegion,
+		}
+		cloudTag, err := names.ParseCloudTag(cons.CloudTag)
+		if err != nil {
+			result[i] = params.InstanceTypesResult{Error: common.ServerError(err)}
+			continue
+		}
+		if m.Cloud() != cloudTag.Id() {
+			result[i] = params.InstanceTypesResult{Error: common.ServerError(errors.NotValidf("asking %s cloud information to %s cloud", cloudTag.Id(), m.Cloud()))}
+			continue
+		}
```

Comment:
I'd check also that cons.CloudRegion is not empty here. otherwise I guess the subsequent error would be ugly and not really informative. This case should also be tested.